File: chapechape_client/lib/core/services/wave_service.dart
import 'package:dio/dio.dart';
import 'package:url_launcher/url_launcher.dart';
import 'dart:async';
import 'package:chapechape_client/core/services/api_service.dart';
import 'package:chapechape_client/core/models/payment_model.dart';

/// Service Wave pour les paiements clients
/// Gère l'initiation des paiements et les redirections vers le portail Wave
class WaveService {
  final ApiService _apiService;
  
  // Configuration URLs
  static const String _successUrl = '/payment-success';
  static const String _cancelUrl = '/payment/cancel';
  
  WaveService._({required ApiService apiService}) : _apiService = apiService;
  
  static Future<WaveService> initialize() async {
    final apiService = await ApiService.initialize();
    return WaveService._(apiService: apiService);
  }

  /// Initier un paiement Wave
  /// Retourne l'URL de redirection vers le portail Wave
  Future<WavePaymentResult> initiatePayment({
    required String reservationId,
    required double amount,
    String? phoneNumber,
    Map<String, dynamic>? metadata,
  }) async {
    try {
      // Appel au backend pour créer l'intention de paiement Wave
      final response = await _apiService.post('/payments/create-payment-intent', data: {
        'reservationId': reservationId,
        'paymentMethod': 'wave',
        'phoneNumber': phoneNumber ?? '', // Utiliser le phoneNumber fourni
      });

      final data = _mapFrom(response.data);
      if (data == null) {
        throw Exception('Réponse API invalide');
      }

      final paymentData = _mapFrom(data['data']);
      final success = data['success'] == true;
      final paymentUrl = _stringFrom(paymentData?['paymentUrl']) ??
          _stringFrom(_mapFrom(paymentData?['providerResponse'])?['paymentUrl']);

      if (success && paymentUrl != null && paymentUrl.isNotEmpty) {
        final provider = _mapFrom(paymentData?['providerResponse']);
        final paymentToken = _stringFrom(paymentData?['paymentToken']) ??
            _stringFrom(provider?['paymentToken']) ??
            _stringFrom(paymentData?['transactionId']);

        DateTime? expiresAt;
        final rawExpiry = paymentData?['expiresAt'];
        if (rawExpiry != null) {
          expiresAt = DateTime.tryParse(rawExpiry.toString());
        }
        expiresAt ??= DateTime.now().add(const Duration(minutes: 30));

        return WavePaymentResult(
          success: true,
          transactionId: _stringFrom(paymentData?['transactionId']),
          paymentUrl: paymentUrl,
          paymentToken: paymentToken,
          expiresAt: expiresAt,
        );
      } else {
        throw Exception(data['message'] ?? 'Erreur lors de l\'initiation du paiement Wave');
      }
      
    } on DioException catch (e) {
      throw WaveException('Erreur réseau: ${e.message}');
    } catch (e) {
      throw WaveException('Erreur lors de l\'initiation du paiement Wave: $e');
    }
  }

  /// Lancer le paiement Wave dans le navigateur externe
  Future<bool> launchPaymentInBrowser(String paymentUrl) async {
    final uri = Uri.parse(paymentUrl);
    if (!uri.hasScheme ||
        (uri.scheme != 'http' && uri.scheme != 'https')) {
      throw WaveException('URL de paiement invalide');
    }
    try {
      // Sur Android 11+, sans <queries> https dans le manifeste, [canLaunchUrl] est souvent false
      // alors qu’un navigateur existe — on tente [launchUrl] directement.
      var launched = await launchUrl(
        uri,
        mode: LaunchMode.externalApplication,
      );
      if (!launched) {
        launched = await launchUrl(uri, mode: LaunchMode.platformDefault);
      }
      if (!launched) {
        throw WaveException('Impossible d\'ouvrir l\'URL de paiement Wave');
      }
      return launched;
    } catch (e) {
      if (e is WaveException) rethrow;
      throw WaveException('Erreur lors du lancement du paiement Wave: $e');
    }
  }

  /// Vérifier le statut d'un paiement Wave via l'endpoint dédié (rapide)
  /// PROB #7 CORRIGÉ : lookup direct, ne charge plus tous les paiements
  Future<WavePaymentStatus> checkPaymentStatus(String transactionId) async {
    try {
      // Endpoint dédié : lookup direct par index MongoDB
      final response = await _apiService.get('/payments/status/$transactionId');

      if (response.data['success'] == true && response.data['payment'] != null) {
        final p = response.data['payment'] as Map<String, dynamic>;
        return WavePaymentStatus(
          transactionId: transactionId,
          status: _mapPaymentStatus(p['status']?.toString()),
          amount: (p['amount'] as num?)?.toDouble(),
          currency: p['currency']?.toString() ?? 'XOF',
          message: p['providerStatus']?.toString() ?? '',
          processedAt: p['updatedAt'] != null
              ? DateTime.tryParse(p['updatedAt'].toString())
              : null,
        );
      }

      return WavePaymentStatus(
        transactionId: transactionId,
        status: PaymentStatus.pending,
        message: 'Paiement non trouvé',
      );
    } on DioException catch (e) {
      throw WaveException('Erreur lors de la vérification du statut Wave: ${e.message}');
    }
  }

  /// Recherche robuste par transactionId (root ou nested)
  Map<String, dynamic>? _findPaymentByTransactionId(String wantedId, List<dynamic> payments) {
    for (final raw in payments.cast<Map<String, dynamic>>()) {
      final txnRoot = raw['transactionId']?.toString();
      final txnNested = raw['paymentDetails']?['providerResponse']?['transactionId']?.toString();
      
      if (wantedId == txnRoot || wantedId == txnNested) {
        print('✅ Match trouvé: ${wantedId == txnRoot ? 'root' : 'nested'} transactionId');
        return raw;
      }
    }
    return null;
  }

  /// Mapper le statut backend vers enum PaymentStatus
  /// PROB #8 CORRIGÉ : centralisé, inclut tous les alias backend
  PaymentStatus _mapPaymentStatus(String? backendStatus) {
    switch (backendStatus?.toLowerCase()) {
      case 'paid':
      case 'completed':
      case 'succeeded':
      case 'success':
        return PaymentStatus.succeeded; // Cohérent avec payment_model.dart
      case 'failed':
      case 'error':
      case 'rejected':
        return PaymentStatus.failed;
      case 'cancelled':
      case 'canceled':
        return PaymentStatus.cancelled;
      case 'processing':
        return PaymentStatus.processing;
      case 'pending':
      default:
        return PaymentStatus.pending;
    }
  }

  /// Valider les paramètres de paiement avant l'initiation
  void _validatePaymentParams({
    required String reservationId,
    required double amount,
  }) {
    if (reservationId.isEmpty) {
      throw WaveException('ID de réservation requis');
    }
    
    if (amount <= 0) {
      throw WaveException('Montant invalide');
    }
    
    // Validation Wave: montant minimum 100 FCFA
    if (amount < 100) {
      throw WaveException('Le montant minimum est de 100 FCFA');
    }
  }

  /// Formater le numéro de téléphone pour Wave
  String formatPhoneNumber(String phoneNumber) {
    // Nettoyer le numéro
    String clean = phoneNumber.replaceAll(RegExp(r'[^\d]'), '');
    
    // Ajouter le préfixe ivoirien si nécessaire
    if (clean.length == 10 && clean.startsWith('0')) {
      clean = '225${clean.substring(1)}';
    } else if (clean.length == 8) {
      clean = '225$clean';
    }
    
    return clean;
  }

  static Map<String, dynamic>? _mapFrom(dynamic value) {
    if (value == null) return null;
    if (value is Map<String, dynamic>) return value;
    if (value is Map) return Map<String, dynamic>.from(value);
    return null;
  }

  static String? _stringFrom(dynamic value) {
    if (value == null) return null;
    final s = value.toString();
    return s.isEmpty ? null : s;
  }
}

/// Exception spécifique Wave
class WaveException implements Exception {
  final String message;
  final String? code;

  WaveException(this.message, [this.code]);

  @override
  String toString() => 'WaveException: $message${code != null ? ' (Code: $code)' : ''}';
}

File: backend/src/services/wave.service.js
const axios = require('axios');
const logger = require('../utils/logger');
const {
    getWavePaymentWebhookSecret,
    verifyWaveWebhookHmac,
} = require('../utils/wave-webhook-signature.util');

/**
 * Service d'intégration Wave pour ChapeChape Residence
 * Support des paiements Wave en Côte d'Ivoire
 */
class WaveService {
    constructor() {
        // Configuration Wave depuis variables d'environnement
        this.apiKey = process.env.WAVE_API_KEY;
        this.baseUrl = process.env.WAVE_BASE_URL || 'https://api.wave.com';
        this.successUrl = process.env.WAVE_SUCCESS_URL || `${process.env.CLIENT_URL}/payment/success`;
        this.errorUrl = process.env.WAVE_ERROR_URL || `${process.env.CLIENT_URL}/payment/error`;
        this.webhookUrl = process.env.WAVE_WEBHOOK_URL || `${process.env.APP_URL}/api/payments/wave/webhook`;
        // Supporte les deux noms de variable d'environnement pour compatibilité
        this.signingSecret = getWavePaymentWebhookSecret();

        // Flag d'activation pour éviter de faire crasher l'app si la config est incomplète
        this.enabled = true;

        // Validation des paramètres obligatoires (sans throw pour ne pas tuer le serveur)
        if (!this.apiKey) {
            logger.error('Wave mal configuré : WAVE_API_KEY manquante');
            this.enabled = false;
        }

        if (!this.baseUrl) {
            logger.error('Wave mal configuré : WAVE_BASE_URL manquante');
            this.baseUrl = 'https://api.wave.com';
        }

        logger.info('Wave Service initialisé', {
            baseUrl: this.baseUrl,
            enabled: this.enabled,
            hasApiKey: !!this.apiKey,
            hasSigningSecret: !!this.signingSecret
        });
    }

    /**
     * Valider les données de paiement
     * @param {Object} paymentData - Données du paiement
     * @param {Object} user - Utilisateur
     * @param {Object} reservation - Réservation
     * @returns {Object} Validation result
     */
    validatePaymentData(paymentData, user, reservation) {
        if (!paymentData.amount || paymentData.amount <= 0) {
            return { valid: false, error: 'Montant invalide' };
        }

        if (!paymentData.phoneNumber) {
            return { valid: false, error: 'Numéro de téléphone requis' };
        }

        if (!user || !user._id) {
            return { valid: false, error: 'Utilisateur invalide' };
        }

        if (!reservation || !reservation._id) {
            return { valid: false, error: 'Réservation invalide' };
        }

        return { valid: true };
    }

    /**
     * Formater le numéro de téléphone pour Wave
     * @param {string} phoneNumber - Numéro brut
     * @returns {string} Numéro formaté
     */
    formatPhoneNumber(phoneNumber) {
        if (!phoneNumber) return null; // Pas de fallback fictif — laisser Wave gérer
        
        // Nettoyer le numéro
        let clean = phoneNumber.replace(/\D/g, '');
        
        // Ajouter le préfixe ivoirien si nécessaire
        if (clean.length === 10 && clean.startsWith('0')) {
            clean = '225' + clean.substring(1);
        } else if (clean.length === 8) {
            clean = '225' + clean;
        }
        
        return '+' + clean;
    }

    // Initialiser un paiement Wave
    async initiatePayment(paymentData, user, reservation) {
        try {
            // Vérifier que le service est correctement configuré
            if (!this.enabled) {
                logger.error('Initiation Wave impossible: service désactivé (configuration manquante)');
                return {
                    success: false,
                    error: 'Service Wave désactivé: configuration manquante',
                    provider: 'wave'
                };
            }

            // Validation des données d'entrée
            const validation = this.validatePaymentData(paymentData, user, reservation);
            if (!validation.valid) {
                logger.error('Validation Wave échouée:', validation.error);
                return {
                    success: false,
                    error: validation.error
                };
            }

            // Construction du payload Wave selon la documentation officielle
            const payload = {
                // Champs obligatoires selon la doc Wave
                amount: paymentData.amount.toString(), // DOIT être string selon la doc
                currency: paymentData.currency || 'XOF',
                success_url: this.successUrl,
                error_url: this.errorUrl,
                
                // Champ optionnel mais recommandé pour traçabilité
                client_reference: reservation._id.toString()
            };

            // Log de la requête (sans apikey pour sécurité)
            logger.info('Initiation paiement Wave', {
                amount: payload.amount,
                currency: payload.currency,
                client_reference: payload.client_reference
            });

            const response = await axios({
                method: 'post',
                url: `${this.baseUrl}/v1/checkout/sessions`, // ✅ Endpoint correct selon la doc
                headers: {
                    'Authorization': `Bearer ${this.apiKey}`,
                    'Content-Type': 'application/json',
                    'User-Agent': 'ChapeChape-Backend/1.0'
                },
                data: payload,
                timeout: 30000 // 30 secondes timeout
            });

            if (response.data && response.data.wave_launch_url) {
                logger.info('Paiement Wave créé avec succès', {
                    transactionId: response.data.id || response.data.session_id,
                    client_reference: payload.client_reference
                });

                return {
                    success: true,
                    transactionId: response.data.id || response.data.session_id,
                    status: 'pending',
                    paymentUrl: response.data.wave_launch_url,
                    paymentToken: response.data.id || response.data.session_id,
                    provider: 'wave',
                    providerResponse: response.data
                };
            } else {
                throw new Error(`Réponse Wave inattendue: ${JSON.stringify(response.data)}`);
            }
        } catch (error) {
            logger.error('Erreur lors de l\'initiation du paiement Wave:', {
                error: error.message,
                reservationId: (reservation && reservation._id) ? reservation._id : (paymentData && paymentData.reservation ? paymentData.reservation : null),
                amount: paymentData && paymentData.amount,
                response: error.response?.data
            });

            // Gestion des erreurs spécifiques Wave
            if (error.response?.data) {
                const waveError = error.response.data;
                return {
                    success: false,
                    error: this.mapWaveError(waveError),
                    errorCode: waveError.code || error.response.status,
                    provider: 'wave'
                };
            }

            return {
                success: false,
                error: error.message || 'Erreur de connexion Wave',
                provider: 'wave'
            };
        }
    }

    /**
     * Mapper les erreurs Wave vers des messages utilisateur
     * @param {Object} waveError - Erreur Wave
     * @returns {string} Message d'erreur utilisateur
     */
    mapWaveError(waveError) {
        const errorMappings = {
            '400': 'Paramètres de paiement invalides',
            '401': 'Clé API Wave invalide',
            '403': 'Accès non autorisé',
            '404': 'Service Wave indisponible',
            '429': 'Trop de requêtes, veuillez réessayer plus tard',
            '500': 'Erreur serveur Wave'
        };

        const errorCode = waveError.code?.toString() || waveError.status?.toString();
        return errorMappings[errorCode] || waveError.message || 'Erreur de paiement Wave inconnue';
    }

    // Vérifier la signature du webhook
    verifySignature(payloadBuffer, signature) {
        if (!this.signingSecret) {
            console.warn('WAVE_SIGNING_SECRET/WAVE_WEBHOOK_SECRET non configuré');
            return false;
        }
        return verifyWaveWebhookHmac(payloadBuffer, signature, this.signingSecret);
    }

    /**
     * Traiter une notification webhook Wave
     * Selon la doc Wave, les webhooks ont des événements checkout.session.*
     * @param {Object} webhookData - Données du webhook
     * @returns {Object} Résultat du traitement
     */
    async processWebhook(webhookData) {
        try {
            logger.info('Traitement webhook Wave:', webhookData);

            // Récupérer l'ID de transaction selon la structure Wave
            const transactionId = webhookData.data?.id || webhookData.id;
            
            if (!transactionId) {
                throw new Error('ID de transaction manquant dans la notification');
            }

            // Déterminer le statut selon les événements Wave documentés
            let status;
            switch(webhookData.event) {
                case 'checkout.session.completed':
                    // Vérifier également payment_status dans les données
                    if (webhookData.data?.payment_status === 'succeeded') {
                        status = 'paid';
                    } else {
                        status = 'pending';
                    }
                    break;
                case 'checkout.session.failed':
                case 'checkout.session.expired':
                    status = 'failed';
                    break;
                default:
                    status = 'pending';
            }

            return {
                success: true,
                transactionId: transactionId,
                status: status,
                webhookData: webhookData
            };
        } catch (error) {
            logger.error('Erreur de traitement webhook Wave:', error);
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * Vérifier le statut d'un paiement Wave
     * @param {string} transactionId - ID de transaction
     * @returns {Object} Statut du paiement
     */
    async checkPaymentStatus(transactionId) {
        try {
            if (!this.enabled) {
                return {
                    success: false,
                    error: 'Service Wave désactivé: configuration manquante'
                };
            }
            if (!transactionId) {
                throw new Error('ID de transaction requis');
            }

            const response = await axios({
                method: 'get',
                url: `${this.baseUrl}/v1/checkout/sessions/${transactionId}`, // ✅ Endpoint correct selon la doc
                headers: {
                    'Authorization': `Bearer ${this.apiKey}`,
                    'Content-Type': 'application/json',
                    'User-Agent': 'ChapeChape-Backend/1.0'
                },
                timeout: 15000
            });

            let status = 'pending';
            // Selon la doc Wave, le statut est dans payment_status, pas status
            const paymentStatus = response.data && response.data.payment_status;
            if (paymentStatus) {
                switch(paymentStatus.toLowerCase()) {
                    case 'succeeded':
                        status = 'paid';
                        break;
                    case 'failed':
                    case 'canceled':
                    case 'cancelled':
                    case 'expired':
                        status = 'failed';
                        break;
                    case 'processing':
                    case 'pending':
                    default:
                        status = 'pending';
                }
            }

            logger.info('Vérification statut Wave', {
                transactionId,
                payment_status: response.data?.payment_status,
                mappedStatus: status
            });

            return {
                success: true,
                status: status,
                data: response.data
            };
        } catch (error) {
            logger.error('Erreur vérification statut Wave:', {
                error: error.message,
                transactionId,
                response: error.response?.data
            });

            return {
                success: false,
                error: error.response?.data?.message || error.message || 'Erreur de vérification statut'
            };
        }
    }
}

module.exports = new WaveService();
